### PR TITLE
master: Disallow duplicate LocalVariableTypeTable entries

### PR DESCRIPTION
**Pull request by theresa-m**
- JVM Spec 4.7.14: There may be no more than one LocalVariableTypeTable attribute per local variable in the attributes table of a Code attribute
- Refactor ClassFormatErrors with messages in ClassFileOracle.cpp

Parts of this are from https://github.com/eclipse/openj9/pull/6289

pr to 0.17.0 release branch: https://github.com/eclipse/openj9/pull/7263

Signed-off-by: Theresa Mammarella <Theresa.T.Mammarella@ibm.com>

### CODE REVIEW

File: runtime/bcutil/ClassFileOracle.cpp [hunk 1]
```diff
@@ -1158,20 +1179,34 @@ ClassFileOracle::walkMethodCodeAttributeAttributes(U_16 methodIndex)
...
 							Trc_BCU_ClassFileOracle_walkMethodCodeAttributeAttributes_DuplicateLocalVariableTypeTable(
-									(U_32)getUTF8Length(_classFile->methods[methodIndex].nameIndex), getUTF8Data(_classFile->methods[methodIndex].nameIndex), localVariableTypeTable, _methodsInfo[methodIndex].localVariablesInfo[index].localVariableTypeTable);
+									(U_32)getUTF8Length(_classFile->methods[methodIndex].nameIndex), getUTF8Data(_classFile->methods[methodIndex].nameIndex), localVariableTypeTable, _methodsInfo[methodIndex].localVariablesInfo[lvttEntry.index].localVariableTypeTable);
 							_buildResult = GenericError;
 							break;
 						}
+
+						/* 4.7.14: There may be no more than one LocalVariableTypeTable attribute per local variable in the attributes table of a Code attribute. */
+						for (U_16 localVariableTypeTableCompareIndex = 0; localVariableTypeTableCompareIndex < localVariableTypeTableIndex; ++localVariableTypeTableCompareIndex) {
+							J9CfrLocalVariableTypeTableEntry lvttCompareEntry = localVariableTypeTable->localVariableTypeTable[localVariableTypeTableCompareIndex];
```
Comment (DanHeidinga) [inline runtime/bcutil/ClassFileOracle.cpp]: This is also a struct copy - intentional?

File: runtime/bcutil/ClassFileOracle.cpp [hunk 2]
```diff
@@ -1158,20 +1179,34 @@ ClassFileOracle::walkMethodCodeAttributeAttributes(U_16 methodIndex)
...
+							_methodsInfo[methodIndex].localVariablesInfo[lvttEntry.index].localVariableTypeTable = localVariableTypeTable;
+						} else if (localVariableTypeTable != _methodsInfo[methodIndex].localVariablesInfo[lvttEntry.index].localVariableTypeTable) {
 							Trc_BCU_ClassFileOracle_walkMethodCodeAttributeAttributes_DuplicateLocalVariableTypeTable(
-									(U_32)getUTF8Length(_classFile->methods[methodIndex].nameIndex), getUTF8Data(_classFile->methods[methodIndex].nameIndex), localVariableTypeTable, _methodsInfo[methodIndex].localVariablesInfo[index].localVariableTypeTable);
+									(U_32)getUTF8Length(_classFile->methods[methodIndex].nameIndex), getUTF8Data(_classFile->methods[methodIndex].nameIndex), localVariableTypeTable, _methodsInfo[methodIndex].localVariablesInfo[lvttEntry.index].localVariableTypeTable);
 							_buildResult = GenericError;
 							break;
 						}
+
+						/* 4.7.14: There may be no more than one LocalVariableTypeTable attribute per local variable in the attributes table of a Code attribute. */
```
[the remark this answers is not in the document]
Comment (theresa-m) [reply runtime/bcutil/ClassFileOracle.cpp]: RI builds Java 8, 11 and 13 all throw `java.lang.ClassFormatError: Duplicated LocalVariableTypeTable attribute entry for 'this'` for this scenario

**Comment by theresa-m** [comment]
Corrected check so that a LocalVariableTypeTable attribute is unique with its starPC, length, and index instead of all five parameters.

File: runtime/bcutil/ClassFileOracle.cpp [hunk 3]
```diff
@@ -1158,20 +1177,33 @@ ClassFileOracle::walkMethodCodeAttributeAttributes(U_16 methodIndex)
 				J9CfrAttributeLocalVariableTypeTable *localVariableTypeTable = (J9CfrAttributeLocalVariableTypeTable *) attribute;
 				if (0 != localVariableTypeTable->localVariableTypeTableLength) {
 					for (U_16 localVariableTypeTableIndex = 0; localVariableTypeTableIndex < localVariableTypeTable->localVariableTypeTableLength; ++localVariableTypeTableIndex) {
-						U_16 index = localVariableTypeTable->localVariableTypeTable[localVariableTypeTableIndex].index;
-						if (codeAttribute->maxLocals <= index) {
+						J9CfrLocalVariableTypeTableEntry *lvttEntry = localVariableTypeTable->localVariableTypeTable + localVariableTypeTableIndex;
```
[the remark this answers is not in the document]
Comment (DanHeidinga) [reply runtime/bcutil/ClassFileOracle.cpp]: This suggestion uses array syntax to get the address of the element and pulls index into a local which simplifies some of the other changes below

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk8

**Comment by DanHeidinga** [comment]
Jenkins test sanity win jdk11

**Comment by DanHeidinga** [comment]
Jenkins test sanity osx jdk13

**Comment by DanHeidinga** [comment]
@theresa-m Can you create a PR against the `v0.17.0-release` branch for this change as well?

**Comment by theresa-m** [comment]
pr for 0.17.0 is here: https://github.com/eclipse/openj9/pull/7263

**Pull request merged**